ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -2089,7 +2100,34 @@ private void finalizeMetadataChunk() {
               passedInBlobProperties.getAccountId(), passedInBlobProperties.getContainerId(),
               passedInBlobProperties.isEncrypted(), passedInBlobProperties.getExternalAssetTag(),
               passedInBlobProperties.getContentEncoding(), passedInBlobProperties.getFilename(), null);
-      if (isStitchOperation() || getNumDataChunks() > 1) {
+
+      if (isS3MultipartPartUpload()) {
```

Comment:
I don't think this is a good idea to have s3 knowledge inside of PutOperation. I would propose another way of convey these information in and out. Basically, we have extra input (a flag to indicate if we want to have composite blob id) and extra output (a list of blob ids in a string). I think we can use PutBlobOptions to pass this information, something PutBlobOption.ignoreCompositeBlobId, and just return a string (maybe in base64) that contains all blob ids and sizes. Something like that.